### DISCUSSION TITLE
Any suggestions on how to speed up reading from a Pyo3 class?

### DISCUSSION THREAD

**Discussion by ollz272** [PyO3/pyo3#3324] [Questions]
Hi!

Love the project - i've been using it for some hobby projects and noticed that, while the speed in the pure rust code is extremely fast, when trying to read/write from python it much slower.

```rs
#[pyclass(subclass)]
pub struct BaseClass {
    #[pyo3(get, set)]
    some_field: Vec<i32>,
    #[pyo3(get, set)]
    some_other_field: Vec<Vec<i32>>,
}
```

```python
class MyClass(BaseClass)
    def do_some_writes(self):
        self.some_field = [
            1,2,3,4,5
        ]
        self.some_other_field = [
            [1,2,3,4,5], [1,2,3,4,5], [1,2,3,4,5]
        ]

   def do_some_reads(self):
        self.some_field
        self.some_other_field
```
Is there anyway I can speed up my read/writes of rust objects from python?

**Comment by adamreichold** [comment]
> Is there anyway I can speed up my read/writes of rust objects from python?

Not generally, but only in certain situations meaning that we need information on how the fields are used.

Generally speaking, crossing the language boundary will always have a cost associated but one can change where and how that cost is paid. For example, you could store `Py<PyList>` instead of `Vec<i32>` inside of `BaseClass` so that setting/getting this value would increment a reference count, but not copy any data. The price to pay for this would be that Rust could now always needs to manually convert the list items to `i32` during processing and has to work with the `PyList` interface instead of `Vec`.
